Feature: Semantic search that helps users search based on meaning using AI

Chamepp opened the pull request:
## Summary
This PR introduces `GPTSemanticSearchAlgorithm`, **a semantic search capability**, into the SearchMind Swift package using the **GPT-based** semantic search algorithm. It adds a new option under `SearchOptions` to enable semantic search and integrates the necessary logic in the _algorithm selector_.

This feature allows developers to perform searches based on **meaning** rather than **keywords**, powered by an _external LLM_ (like OpenAI).

## Changes
- **Semantic Search Added**
   - Added a new boolean flag semantic in `SearchOptions`
   - Since `PatternMatchAlgorithm` was the only algorithm for `fileContents` search, it was true by default when it comes to searching for fileContents. Defined the `PatternMatchAlgorithm` to false, since we are adding another search algorithm that serves this purpose.
   - Updated `SearchAlgorithmSelector` to return `GPTSemanticSearchAlgorithm` when `.semantic` is enabled
   - Connected semantic search with API key lookup using `ProcessInfo.processInfo.environment`
       - Methods for securing `API KEYS` in development such as _.env, .xcconfig or Info.plist_ were not a great fit for us since they .env is not suited for the XCode environment and the other two methods were for app development.
       - The current method uses the processInfo to read the environment variable of the `OPENAI_API_KEY` earlier added
       - Before testing the swift package this environment variable should be added by the `export OPENAI_API_KEY="API_KEY"` command, otherwise the tests will fail due to the fact that the api key is missing.
    - Implemented error handling when API key is missing or decoding fails

- **Documentation (DocC)**
   - Documented **semantic** and **patternMatch** options in `SearchOptions`
   - Added performance and error-related notes for **patternMatch** (regex-based)

- **Tests**
   - Initial test case for `testSemanticSearch` added

## Next Steps
- Handling the api response errors in a better way by leveraging the current `SearchError.swift`.
- Expanding testSemanticSearch() further by adding more files in the test setup that help with a more in depth testing for the embedding and relevance scores.
- Improving the `cosineSimilarity` function using the swift-numerics package already available, if it benefits in performance terms. If not, we should avoid this because we add _unnecessary complexity_.
- Asking the developer for their api key in usage and take it as a **parameter** for the function
- Getting the **api key as a parameter** when the semantic search algorithm is selected. This data can be added as a _config_, which can be a struct that holds other data as well that can be used for _more customization_ of the experience.
   - This can be further scaled to the point that other _search algorithms_ can benefit from this too.
   - An idea that comes to my mind, which is not yet validated is to give the user the ability to select the algorithm by using an enum struct, and then getting a config struct which holds preferences like case _sensitivity, api key, api endpoint and etc._ By doing this we can benefit from the `Strategy Design Pattern`.
- Adding the possibility to use **different api endpoints**, other than OpenAI
   - For now we added _OpenAI endpoint_ which is the best and the most popular option among developers
   - The api endpoint `can be set alongside the api key` in the `config struct` mentioned earlier
- By following the rule of three, the `getFilesToSearch` function can be refactored into a single function that acts as a helper in the `Utils` folder that can be located in the internal folder.
- By following the `SRP principle`, algorithms can be break down into their own related file that follow a single responsibility.
- We considered enabling offline semantic embedding calculation by allowing `.coreml AI` models to be embedded within the project.
   - However, converted models (e.g., from OpenAI or Sentence Transformers) typically **consume unnecessarily large space (e.g., 64MB or more)**, which makes bundling them in the package unreasonable.
   - Instead, we can offer a hook for developers to provide their own Core ML model for embeddings if one already exists in their app. While this would **enable fully offline semantic search**, we consider it a non-priority luxury feature for now.
   - Supporting offline semantic search would be useful for **system-wide local searches**, but it becomes less relevant when the goal is to support semantic search over remote databases, which require internet connectivity.
   - That said, we recognize there are cases where **local databases (e.g., Core Data or SQLite)** are used, and may themselves be offline (disconnected from cloud sync, or fully local-first), so future support for offline embeddings and search could be revisited based on developer demand.

froggomad commented on Sources/SearchMind/Internal/SearchEngine.swift:
```diff
@@ -466,4 +470,134 @@ internal struct PatternMatchAlgorithm: SearchAlgorithm {
+
+    private func embed(text: String) async throws -> [Double] {
+        guard let apiKey = ProcessInfo.processInfo.environment["OPENAI_API_KEY"] else {
+                  throw NSError(domain: "OpenAI", code: 0, userInfo: [NSLocalizedDescriptionKey: "API key not found in .env"])
```
This is great that you understand `NSError` and how to use it. Are you familiar with custom Errors? It's quite easy in Swift and can be extended easily as well.

First, you create an enum. In this case, I might call it `enum OpenAIError`. I would add `Error` conformance to the enum and give it 2 cases - `case embedding` and `case APIKeyNotFound`.

```swift
enum OpenAIError: Error {
    case APIKeyNotFound
    case embeddingError
}
```

Then when I want to throw an error, I can simply call `throw OpenAIError.APIKeyNotFound`. In an application for a large audience of developers, I might even go a step further and conform the enum to [LocalizedError](https://developer.apple.com/documentation/foundation/localizederror) where I can provide debugging information at the call site by downcasting the thrown error to `LocalizedError` and use properties such as `failureReason` and `recoverySuggestion`.

That might look like this (I didn't run this through Xcode so it may not compile as-is):

```swift
enum OpenAIError: LocalizedError {
    case APIKeyNotFound
    case embeddingError
    
    var failureReason: String {
        switch self {
            case APIKeyNotFound:
                "Your OpenAI API Key is missing."
             ...          
        }
    }
    
    var recoverySuggestion: String {
         switch swlf {
             case APIKeyNotFound:
                 "Provide one by inserting it as an environment variable in your active scheme."
              ...
         }
    }
}
```

Chamepp replied:
Yes you teached me before on how i can implement custom errors, thank you for your explanation. I was planning to implement this in the current `SearchError.swift` file, but I wanted to push and avoid working on perfection. my next step is to add new errors for api response in the `SearchError.swift`

froggomad commented on Sources/SearchMind/Internal/SearchEngine.swift:
```diff
@@ -466,4 +470,134 @@ internal struct PatternMatchAlgorithm: SearchAlgorithm {
+        guard let apiKey = ProcessInfo.processInfo.environment["OPENAI_API_KEY"] else {
+                  throw NSError(domain: "OpenAI", code: 0, userInfo: [NSLocalizedDescriptionKey: "API key not found in .env"])
+              }
+        let request = EmbeddingRequest(model: "text-embedding-ada-002", input: [text])
```
You might consider allowing developers to use different embedding models to fit their needs and budget. To do this, you can add a parameter `embeddingModel: String` to the function signature. You can then give it a default of "text-embedding-ada-002" to make it easy for those who don't care to customize. That way, those who aren't familiar with the pros and cons of using different embedding models have the choice made for them, and those who do care are able to override the default.

```swift
func search(term: String, type: SearchType, options: SearchOptions, embeddingModel: String = "text-embedding-ada-002") async throws -> [SearchMind.SearchResult]
```

Chamepp replied:
This struct can even contain more data for searching preferences.

froggomad commented on Sources/SearchMind/Internal/SearchEngine.swift:
```diff
@@ -74,10 +74,14 @@ internal final class SearchAlgorithmSelector: Sendable {
         }
-        
+
+        if type == .fileContents && options.semantic {
```
be mindful that if options.patternMatch and options.semantic are both true, you pick one or the other. Right now, semantic overrides pattern-based. That might be exactly what you want, or you might need a more explicit decision logic. I'm a fan of Railway Oriented Programming, but it can be a bit complex. I like it because it doesn't leave side-effects as a possibility. Take this scenario as an example where a developer passed `patternMatch` and `semantic`. They might expect both searches to take place, but only the semantic search is used.

Assuming this is the intent, it's often best to make the route of passing in both options impossible. https://medium.com/@pavelshilyagov/railway-oriented-programming-in-swift-d4131c05d627

Chamepp replied:
Wow that's really interesting, I haven’t heard about this before, sounds like a great solution to this. I will research more on this and open a PR for solving the algorithm selection issue after I submitted the database feature 🙏

Chamepp replied:
Yes I believe the same. Its getting a bit confusing as the options are rising.

froggomad commented on Sources/SearchMind/Internal/SearchEngine.swift:
```diff
@@ -466,4 +470,134 @@ internal struct PatternMatchAlgorithm: SearchAlgorithm {
+        let searchTerm = options.caseSensitive ? term : term.lowercased()
+        let termEmbedding = try await embed(text: searchTerm)
+
+        let files = try await getFilesToSearch(options: options)
```
This is a bit advanced, but for large directories, it may be best to handle this concurrently with a `TaskGroup`. This allows each file's embedding to be fetched in parallel rather than waiting serially for each one to complete. However, one must then consider rate limits - OpenAI may not allow so many searches in parallel.

```swift
let files = try await getFilesToSearch(options: options)
let searchTermEmbedding = try await embed(text: searchTerm)

let results = await withTaskGroup(of: SearchResult?.self) { group in
    for fileURL in files {
        group.addTask {
            do {
                let content = try String(contentsOf: fileURL)
                let fileEmbedding = try await embed(text: content)
                let similarity = cosineSimilarity(searchTermEmbedding, fileEmbedding)
                if similarity > 0.4 {
                    return SearchMind.SearchResult(
                        matchType: .fileContents,
                        path: fileURL.path,
                        relevanceScore: similarity,
                        matchedTerms: [term]
                    )
                }
            } catch {
                // Handle or log error, returning nil
            }
            return nil
        }
    }
    
    var partialResults: [SearchMind.SearchResult] = []
    for await maybeResult in group {
        if let result = maybeResult { partialResults.append(result) }
    }
    return partialResults
}

return results
    .sorted { $0.relevanceScore > $1.relevanceScore }
    .prefix(options.maxResults)
    .map { $0 }
```

froggomad reviewed (approved):
@Chamepp This is a fantastic solution to semantic search! Your code is very clean for the most part. I haven't had a chance to pull it down and test locally yet, but from my analysis, it looks like it will.

My main suggestions are quality of life improvements.

Chamepp commented:
@froggomad I’ve created a few issues addressing the points you mentioned. I’d appreciate it if you could take a look and share your thoughts whenever you had time.
